src/main/java/com/squareup/javapoet/ParameterizedTypeName.java

```diff
@@ -37,37 +42,101 @@
 
   ParameterizedTypeName(ClassName rawType, List<TypeName> typeArguments,
       List<AnnotationSpec> annotations) {
+    this(
+        rawType,
+        rawType.enclosingClassName() != null ? rawType.enclosingClassName() : rawType,
+        rawType.enclosingClassName() != null ? rawType.simpleName() : null,
+        typeArguments,
+        annotations);
+  }
+
+  private ParameterizedTypeName(
+      ClassName rawType,
+      TypeName enclosingTypeOrRawType,
+      String nestedClassName,
+      List<TypeName> typeArguments,
+      List<AnnotationSpec> annotations) {
     super(annotations);
     this.rawType = checkNotNull(rawType, "rawType == null");
+    this.enclosingTypeOrRawType =
+        checkNotNull(enclosingTypeOrRawType, "enclosingTypeOrRawType == null");
+    this.nestedClassName = nestedClassName;
     this.typeArguments = Util.immutableList(typeArguments);
 
-    checkArgument(!this.typeArguments.isEmpty(), "no type arguments: %s", rawType);
+    checkArgument(
+        !this.typeArguments.isEmpty() || enclosingTypeOrRawType instanceof ParameterizedTypeName,
+        "no type arguments: %s",
+        enclosingTypeOrRawType);
     for (TypeName typeArgument : this.typeArguments) {
       checkArgument(!typeArgument.isPrimitive() && typeArgument != VOID,
           "invalid type parameter: %s", typeArgument);
     }
   }
 
   @Override public ParameterizedTypeName annotated(List<AnnotationSpec> annotations) {
-    return new ParameterizedTypeName(rawType, typeArguments, concatAnnotations(annotations));
+    return new ParameterizedTypeName(
```

Comment:
(the rest of this codebase doesn’t generally do line-per-item arguments or parameter declarations)